### PR TITLE
add processing "cache" and augmentation

### PR DESCRIPTION
**Pull request by flozi00**
# What does this PR do?
This PR stores the resampled commoinvoice on disk to speedup multiple runs.
Furthermore it adds data augmentation to double the dataset size

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@patrickvonplaten @patil-suraj 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: examples/research_projects/wav2vec2/run_common_voice.py [hunk 1]
```diff
@@ -407,6 +417,38 @@ def speech_file_to_array_fn(batch):
         remove_columns=eval_dataset.column_names,
         num_proc=data_args.preprocessing_num_workers,
     )
+    
+    
+    # augmenting the datasets.
+    # We need to read the aduio files as arrays, minupulate the audio randomly and tokenize the targets.
+    augment = Compose([
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_common_voice.py:424]: This should be done optionally IMO and not be the default -> could we add an argument `args.augment_speech_data`?

File: examples/research_projects/wav2vec2/run_common_voice.py [hunk 2]
```diff
@@ -391,11 +394,18 @@ def extract_all_chars(batch):
...
     # We need to read the aduio files as arrays and tokenize the targets.
     def speech_file_to_array_fn(batch):
-        speech_array, sampling_rate = torchaudio.load(batch["path"])
-        batch["speech"] = resampler(speech_array).squeeze().numpy()
+        try:
+            speech_array, sampling_rate = sf.read(batch["path"] + ".wav")
+        except:
+            speech_array, sampling_rate = torchaudio.load(batch["path"])
+            speech_array = resampler(speech_array).squeeze().numpy()
+            sf.write(batch["path"] + ".wav", speech_array, sampling_rate, subtype='PCM_24')
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_common_voice.py:402]: I'm not in favor of writing the processed file to the disk. When applying `.map(...)` the output of this function is already cached

File: examples/research_projects/wav2vec2/run_common_voice.py [hunk 3]
```diff
@@ -407,6 +417,38 @@ def speech_file_to_array_fn(batch):
...
+    
+    
+    augmented_dataset = augmented_dataset.map(
+        augment_speech_file_to_array_fn,
+        remove_columns=train_dataset.column_names,
+        num_proc=data_args.preprocessing_num_workers,
+    )
+    
+    #we need to concatenate the augmented dataset and the raw dataset
+    train_dataset = concatenate_datasets([train_dataset, augmented_dataset])
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_common_voice.py:450]: very nice!

File: examples/research_projects/wav2vec2/run_common_voice.py [hunk 4]
```diff
@@ -391,11 +394,18 @@ def extract_all_chars(batch):
     # Preprocessing the datasets.
     # We need to read the aduio files as arrays and tokenize the targets.
     def speech_file_to_array_fn(batch):
-        speech_array, sampling_rate = torchaudio.load(batch["path"])
-        batch["speech"] = resampler(speech_array).squeeze().numpy()
+        try:
+            speech_array, sampling_rate = sf.read(batch["path"] + ".wav")
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_common_voice.py:398]: is `sf.read` faster in your experiments? Also Common Voice is stored as `.mp3` I think -> don't think `soundfile` can read this no?

**Review by patrickvonplaten — COMMENTED** [review]
Hey @flozi00,

Thanks a lot for your PR - this looks like a very cool addition! Also, it's great to see how few lines of code are needed for the data augmentation part

A couple of points:

- Could we make data augmentation optionally and not force the user to run it every time? IMO, the default should be "false" to not break backwards compatibility
- I'm not a fan of writing the processed datasets to disk by default -> disk space is also limited and the `.map(...)` already caches its output

File: examples/research_projects/wav2vec2/run_common_voice.py [hunk 5]
```diff
@@ -407,6 +417,38 @@ def speech_file_to_array_fn(batch):
...
+    
+    
+    # augmenting the datasets.
+    # We need to read the aduio files as arrays, minupulate the audio randomly and tokenize the targets.
+    augment = Compose([
+        AddGaussianNoise(min_amplitude=0.0001, max_amplitude=0.001, p=0.8),
+        PitchShift(min_semitones=-1, max_semitones=1, p=0.8),
+        Gain(min_gain_in_db=-6, max_gain_in_db=6, p=0.8)
+    ])
+    def augment_speech_file_to_array_fn(batch):
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_common_voice.py:429]: nice!

File: examples/research_projects/wav2vec2/run_common_voice.py (hunk 2 above)
Comment (flozi00) [reply examples/research_projects/wav2vec2/run_common_voice.py:402]: Yeah, but in my experiments it's not working after changing parameters, like filtering etc.
Sometimes I use only 10% after breaking changes for testing purposes and when using the full dataset it's not cached anymore and needs to be resampled again.
Maybe as optional argument to if audio should be stored on disk ?

File: examples/research_projects/wav2vec2/run_common_voice.py (hunk 4 above)
Comment (flozi00) [reply examples/research_projects/wav2vec2/run_common_voice.py:398]: Yes, in my Benchmarks soundfile is the fastest method to load audio, but only wav.
That's why I store the audio as wav after resampling

File: examples/research_projects/wav2vec2/run_common_voice.py (hunk 1 above)
Comment (flozi00) [reply examples/research_projects/wav2vec2/run_common_voice.py:424]: Yes
What about the mass of augmented data ?
Maybe another argument how much percentage should be resampled ?

Comment (patrickvonplaten) [reply examples/research_projects/wav2vec2/run_common_voice.py:424]: I'm fine with that!

File: examples/research_projects/wav2vec2/run_common_voice.py (hunk 2 above)
Comment (patrickvonplaten) [reply examples/research_projects/wav2vec2/run_common_voice.py:402]: Puuh still not really in favor of it. I don't think a 10% speed-up is worth it here to write both the map function and the soundfiles to cache, but ok to do it with a flag! Let's default  the flag to False then :-)

**Comment by patrickvonplaten** [comment]
Hey @flozi00,

Could you make the code quality test pass -> then I think we can merge this one :-)

**Pull request closed without merging**